src/Microsoft.AspNet.OData.Shared/Query/SkipTokenQueryOption.cs

```diff
@@ -0,0 +1,352 @@
+                }
+                else
+                {
+                    value = lastMember.GetType().GetProperty(property.Name).GetValue(lastMember);
+                }
+
+                String uriLiteral = String.Empty;
+                if (TypeHelper.IsEnum(value.GetType()))
+                {
+                    ODataEnumValue enumValue = new ODataEnumValue(value.ToString(), value.GetType().FullName);
+                    uriLiteral = ODataUriUtils.ConvertToUriLiteral(enumValue, ODataVersion.V401, model);
+                }
+                else
+                {
+                    uriLiteral = ODataUriUtils.ConvertToUriLiteral(value, ODataVersion.V401, model);
+                }
+                skipTokenvalue += property.Name + ":" + uriLiteral + (islast ? String.Empty : ",");
+                count++;
+            }
+
+            return skipTokenvalue;
+        }
+
+        private static IEnumerable<IEdmProperty> GetPropertiesForSkipToken(object lastMember, IEdmModel model, OrderByQueryOption orderByQueryOption)
+        {
+            IEdmType edmType = GetTypeFromObject(lastMember, model);
+            IEdmEntityType entity = edmType as IEdmEntityType;
+            if (entity == null)
+            {
+                return null;
+            }
+
+            IList<IEdmProperty> key = entity.Key().AsIList<IEdmProperty>();
+            if (orderByQueryOption != null)
+            {
+                IList<IEdmProperty> orderByProps = orderByQueryOption.OrderByNodes.OfType<OrderByPropertyNode>().Select(p => p.Property).AsIList();
+                foreach (IEdmProperty subKey in key)
+                {
+                    orderByProps.Add(subKey);
+                }
+
+                return orderByProps.AsEnumerable();
+            }
+            return key.AsEnumerable();
+        }
+
+        private static IEdmType GetTypeFromObject(object obj, IEdmModel model)
+        {
+            SelectExpandWrapper selectExpand = obj as SelectExpandWrapper;
+            if (selectExpand != null)
+            {
+                return selectExpand.GetEdmType(model);
+            }
+
+            Type clrType = obj.GetType();
+            return EdmLibHelpers.GetEdmType(model, clrType);
+        }
+
+        private static object FetchLast(IQueryable queryable)
+        {
+            MethodInfo genericMethod = _fetchLastItemInResults.MakeGenericMethod(queryable.ElementType);
+            object[] args = new object[] { queryable };
+            object results = genericMethod.Invoke(null, args);
+            return results;
+        }
+
+        /// <summary>
+        /// XML doc, make it public to avoid security issues
+        /// </summary>
+        /// <typeparam name="T"></typeparam>
+        /// <param name="records"></param>
+        /// <returns></returns>
+        public static T FetchLastItemInResults<T>(IQueryable<T> records)
+        {
+            T[] arr = records.AsArray();
+            return arr[arr.Length - 1];
```

Comment:
Can you use records.LastOrDefault(), so that you don't have to materialize the entire result as an array?

Reply:
Changed the approach.